File: problems/tier1-foundation/008-file-search/solution.js
'use strict';

// File Search — Strategy + Composite + Decorator reference solution (JavaScript).

class FileNode {
  constructor(name, size, extension, isDirectory, children = null) {
    this.name = name;
    this.size = size;
    this.extension = extension;
    this.isDirectory = isDirectory;
    this.children = children ? children : [];
  }
}

class SearchByExtension {
  constructor(ext) { this.ext = ext; }
  matches(file) { return !file.isDirectory && file.extension === this.ext; }
}

class SearchByMinSize {
  constructor(minSize) { this.minSize = minSize; }
  matches(file) { return !file.isDirectory && file.size >= this.minSize; }
}

class SearchByName {
  constructor(substring) { this.substring = substring; }
  matches(file) { return file.name.includes(this.substring); }
}

class AndFilter {
  constructor(criteria) { this.criteria = criteria; }
  matches(file) { return this.criteria.every(c => c.matches(file)); }
}

class OrFilter {
  constructor(criteria) { this.criteria = criteria; }
  matches(file) { return this.criteria.some(c => c.matches(file)); }
}

function _dfs(node, criteria, results) {
  if (node === null || node === undefined) return;
  if (!node.isDirectory && criteria.matches(node)) results.push(node);
  for (const c of node.children) _dfs(c, criteria, results);
}

function search_by_extension(root, ext) {
  const out = [];
  _dfs(root, new SearchByExtension(ext), out);
  return out;
}

function search_by_size(root, minSize) {
  const out = [];
  _dfs(root, new SearchByMinSize(minSize), out);
  return out;
}

function search_by_name(root, substring) {
  const out = [];
  _dfs(root, new SearchByName(substring), out);
  return out;
}

function search_composite(root, criteria, mode) {
  const f = mode === 'AND' ? new AndFilter(criteria) : new OrFilter(criteria);
  const out = [];
  _dfs(root, f, out);
  return out;
}

function search_and_sort(root, criteria, sortBy) {
  const out = [];
  _dfs(root, criteria, out);
  if (sortBy === 'name') {
    out.sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
  } else if (sortBy === 'size') {
    out.sort((a, b) => b.size - a.size);
  } else if (sortBy === 'extension') {
    out.sort((a, b) => (a.extension < b.extension ? -1 : a.extension > b.extension ? 1 : 0));
  }
  return out;
}

// ─── Module fixture state ───────────────────────────────────────────────────

let _g_root = null;

function reset_service() {
  _g_root = null;
}

function fs_build_default_tree() {
  const main_cpp = new FileNode('main.cpp', 50, 'cpp', false);
  const utils_cpp = new FileNode('utils.cpp', 120, 'cpp', false);
  const helper_h = new FileNode('helper.h', 10, 'h', false);
  const readme = new FileNode('readme.md', 5, 'md', false);
  const report = new FileNode('report.pdf', 200, 'pdf', false);
  const build_sh = new FileNode('build.sh', 2, 'sh', false);
  const src = new FileNode('src', 0, '', true, [main_cpp, utils_cpp, helper_h]);
  const docs = new FileNode('docs', 0, '', true, [readme, report]);
  _g_root = new FileNode('project', 0, '', true, [src, docs, build_sh]);
}

function fs_build_empty_tree() {
  _g_root = new FileNode('empty', 0, '', true);
}

function fs_build_single_file_tree() {
  const f = new FileNode('test.txt', 30, 'txt', false);
  _g_root = new FileNode('root', 0, '', true, [f]);
}

function fs_count_by_extension(ext) {
  return _g_root ? search_by_extension(_g_root, ext).length : 0;
}

function fs_has_by_extension(ext, name) {
  if (_g_root === null) return false;
  return search_by_extension(_g_root, ext).some(f => f.name === name);
}

function fs_count_by_size(minSize) {
  return _g_root ? search_by_size(_g_root, minSize).length : 0;
}

function fs_has_by_size(minSize, name) {
  if (_g_root === null) return false;
  return search_by_size(_g_root, minSize).some(f => f.name === name);
}

function fs_count_by_name(sub) {
  return _g_root ? search_by_name(_g_root, sub).length : 0;
}

function fs_has_by_name(sub, name) {
  if (_g_root === null) return false;
  return search_by_name(_g_root, sub).some(f => f.name === name);
}

function fs_count_composite_and(ext, minSize) {
  if (_g_root === null) return 0;
  return search_composite(_g_root, [new SearchByExtension(ext), new SearchByMinSize(minSize)], 'AND').length;
}

function fs_count_composite_or(ext, minSize) {
  if (_g_root === null) return 0;
  return search_composite(_g_root, [new SearchByExtension(ext), new SearchByMinSize(minSize)], 'OR').length;
}

function fs_first_sorted_by(ext, sortBy) {
  if (_g_root === null) return '';
  const v = search_and_sort(_g_root, new SearchByExtension(ext), sortBy);
  return v.length ? v[0].name : '';
}

module.exports = {
  FileNode,
  SearchByExtension,
  SearchByMinSize,
  SearchByName,
  AndFilter,
  OrFilter,
  search_by_extension,
  search_by_size,
  search_by_name,
  search_composite,
  search_and_sort,
  reset_service,
  fs_build_default_tree,
  fs_build_empty_tree,
  fs_build_single_file_tree,
  fs_count_by_extension,
  fs_has_by_extension,
  fs_count_by_size,
  fs_has_by_size,
  fs_count_by_name,
  fs_has_by_name,
  fs_count_composite_and,
  fs_count_composite_or,
  fs_first_sorted_by,
};

File: problems/tier1-foundation/008-file-search/solution.java
// File Search — Solution (Java, Strategy + Composite)
import java.util.*;

class FileNode {
    public String name;
    public int size;
    public String extension;
    public boolean isDirectory;
    public List<FileNode> children;

    public FileNode(String name, int size, String extension, boolean isDirectory) {
        this(name, size, extension, isDirectory, new ArrayList<>());
    }

    public FileNode(String name, int size, String extension, boolean isDirectory, List<FileNode> children) {
        this.name = name; this.size = size; this.extension = extension;
        this.isDirectory = isDirectory; this.children = children;
    }
}

interface SearchCriteria {
    boolean matches(FileNode file);
}

class SearchByExtension implements SearchCriteria {
    private final String ext;
    public SearchByExtension(String ext) { this.ext = ext; }
    public boolean matches(FileNode file) {
        return !file.isDirectory && Objects.equals(file.extension, ext);
    }
}

class SearchByMinSize implements SearchCriteria {
    private final int minSize;
    public SearchByMinSize(int minSize) { this.minSize = minSize; }
    public boolean matches(FileNode file) {
        return !file.isDirectory && file.size >= minSize;
    }
}

class SearchByName implements SearchCriteria {
    private final String substring;
    public SearchByName(String substring) { this.substring = substring; }
    public boolean matches(FileNode file) {
        return file.name.contains(substring);
    }
}

class AndFilter implements SearchCriteria {
    private final List<SearchCriteria> criteria;
    public AndFilter(List<SearchCriteria> criteria) { this.criteria = criteria; }
    public boolean matches(FileNode file) {
        for (SearchCriteria c : criteria) if (!c.matches(file)) return false;
        return true;
    }
}

class OrFilter implements SearchCriteria {
    private final List<SearchCriteria> criteria;
    public OrFilter(List<SearchCriteria> criteria) { this.criteria = criteria; }
    public boolean matches(FileNode file) {
        for (SearchCriteria c : criteria) if (c.matches(file)) return true;
        return false;
    }
}

public class Solution {
    private static FileNode root = null;

    private static void dfs(FileNode node, SearchCriteria c, List<FileNode> results) {
        if (node == null) return;
        if (!node.isDirectory && c.matches(node)) results.add(node);
        for (FileNode ch : node.children) dfs(ch, c, results);
    }

    private static List<FileNode> searchByExtension(FileNode root, String ext) {
        List<FileNode> out = new ArrayList<>();
        dfs(root, new SearchByExtension(ext), out);
        return out;
    }

    private static List<FileNode> searchBySize(FileNode root, int minSize) {
        List<FileNode> out = new ArrayList<>();
        dfs(root, new SearchByMinSize(minSize), out);
        return out;
    }

    private static List<FileNode> searchByName(FileNode root, String sub) {
        List<FileNode> out = new ArrayList<>();
        dfs(root, new SearchByName(sub), out);
        return out;
    }

    public static void reset_service() { root = null; }

    public static void fs_build_default_tree() {
        FileNode mainCpp = new FileNode("main.cpp", 50, "cpp", false);
        FileNode utilsCpp = new FileNode("utils.cpp", 120, "cpp", false);
        FileNode helperH = new FileNode("helper.h", 10, "h", false);
        FileNode readme = new FileNode("readme.md", 5, "md", false);
        FileNode report = new FileNode("report.pdf", 200, "pdf", false);
        FileNode buildSh = new FileNode("build.sh", 2, "sh", false);
        FileNode src = new FileNode("src", 0, "", true,
                new ArrayList<>(Arrays.asList(mainCpp, utilsCpp, helperH)));
        FileNode docs = new FileNode("docs", 0, "", true,
                new ArrayList<>(Arrays.asList(readme, report)));
        root = new FileNode("project", 0, "", true,
                new ArrayList<>(Arrays.asList(src, docs, buildSh)));
    }

    public static void fs_build_empty_tree() {
        root = new FileNode("empty", 0, "", true);
    }

    public static void fs_build_single_file_tree() {
        FileNode f = new FileNode("test.txt", 30, "txt", false);
        root = new FileNode("root", 0, "", true,
                new ArrayList<>(Arrays.asList(f)));
    }

    public static int fs_count_by_extension(String ext) {
        return root == null ? 0 : searchByExtension(root, ext).size();
    }

    public static boolean fs_has_by_extension(String ext, String name) {
        if (root == null) return false;
        for (FileNode f : searchByExtension(root, ext)) if (f.name.equals(name)) return true;
        return false;
    }

    public static int fs_count_by_size(int minSize) {
        return root == null ? 0 : searchBySize(root, minSize).size();
    }

    public static boolean fs_has_by_size(int minSize, String name) {
        if (root == null) return false;
        for (FileNode f : searchBySize(root, minSize)) if (f.name.equals(name)) return true;
        return false;
    }

    public static int fs_count_by_name(String sub) {
        return root == null ? 0 : searchByName(root, sub).size();
    }

    public static boolean fs_has_by_name(String sub, String name) {
        if (root == null) return false;
        for (FileNode f : searchByName(root, sub)) if (f.name.equals(name)) return true;
        return false;
    }

    public static int fs_count_composite_and(String ext, int minSize) {
        if (root == null) return 0;
        List<SearchCriteria> criteria = new ArrayList<>();
        criteria.add(new SearchByExtension(ext));
        criteria.add(new SearchByMinSize(minSize));
        List<FileNode> out = new ArrayList<>();
        dfs(root, new AndFilter(criteria), out);
        return out.size();
    }

    public static int fs_count_composite_or(String ext, int minSize) {
        if (root == null) return 0;
        List<SearchCriteria> criteria = new ArrayList<>();
        criteria.add(new SearchByExtension(ext));
        criteria.add(new SearchByMinSize(minSize));
        List<FileNode> out = new ArrayList<>();
        dfs(root, new OrFilter(criteria), out);
        return out.size();
    }

    public static String fs_first_sorted_by(String ext, String sortBy) {
        if (root == null) return "";
        List<FileNode> v = new ArrayList<>();
        dfs(root, new SearchByExtension(ext), v);
        if (sortBy.equals("name")) v.sort(Comparator.comparing(f -> f.name));
        else if (sortBy.equals("size")) v.sort((a, b) -> Integer.compare(b.size, a.size));
        else if (sortBy.equals("extension")) v.sort(Comparator.comparing(f -> f.extension));
        return v.isEmpty() ? "" : v.get(0).name;
    }
}
